src/CacheTower.Extensions.Redis/RedisLockExtension.cs

```diff
@@ -87,7 +88,13 @@ public async ValueTask<CacheEntry<T>> WithRefreshAsync<T>(string cacheKey, Func<
 			}
 			else
 			{
-				var completionSource = LockedOnKeyRefresh.GetOrAdd(cacheKey, key => new TaskCompletionSource<bool>());
+				var completionSource = LockedOnKeyRefresh.GetOrAdd(cacheKey, key =>
+				{
+					var tcs = new TaskCompletionSource<bool>();
+					var cts = new CancellationTokenSource(Options.LockTimeout);
+					cts.Token.Register(() => tcs.TrySetCanceled(), useSynchronizationContext: false);
```

Comment:
```suggestion
					cts.Token.Register(tcs => ((TaskCompletionSource<bool>)tcs).TrySetCanceled(), tcs, useSynchronizationContext: false);
```

This helps avoid some allocations by avoiding a closure. There is technically a second optimization by switching to `TaskCompletionSource` (the non-generic one) too but that is out-of-scope of this issue.

Reply:
Yeah, I did notice we don't use the return value of that. I figured a task cancelled thrown up the stack was the only outcome that makes sense in this case. Ah those tricky closure allocations! Thanks for the suggestion, added to the PR